### DISCUSSION TITLE
How to pass variable value from child component which is client component to a parent component which is server component without making the parent element a client componnet?

### DISCUSSION THREAD

**Discussion by surafel-99** [vercel/next.js#65012] [Help] [answered]
### Summary

In my client component I have a toggle button that uses useStates and toggles a variable value between True and False. And I wanted to pass the variable value to the parent element.

And when I tied to run the code I got an error saying:

"Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with 'use server' "

Please Help!

### page.tsx

```js
//page.tsx

export default function Home() {

  const getToggleValue = (value) => {
  
    console.log (value);    // I was trying to read the passed value here
  }

  return (
    <main className="">
      <div className="flex flex-col items-center">
          <SearchBar />
      </div>

      <div className="flex justify-end mr-[50px]">
           <ToggleFilter toggleValue={getToggleValue} />
      </div>
      
```
### ToggleFilter.tsx

```js
      // ToggleFilter.tsx 
      
      const ToggleFilter = ({toggleValue} : any) => {

      const [showDescription, updateShowDescription] = useState(false);
    
      const toggleDescription = () => {
        if (showDescription === false) {
          updateShowDescription(true);
        }
        else {
          updateShowDescription(false)
        }
      }
      
      toggleValue(showDescription);     //This is me trying to pass the updated  value to the parent element lol

  return (
    <div>
        <button 
        className="py-2 px-4 bg-sky-600"
        onClick={toggleDescription}>
            <LuShrink className="text-[22px]"/>
        </button>
    </div>
  )
}

export default ToggleFilter
```

**Comment by icyJoseph** [comment, marked as answer]
Yeah, for this one you'd need to make a server action. Since, `getToggleValue`, lives in a server component, you can just do:

```js
 // it has to be async!
 const getToggleValue = async (value) => {
    "use server";
    console.log (value);    // I was trying to read the passed value here
  }
 ```

Though, do remember that the Server Component, only ever exists in the server, and once it pushes out its serialized VDOM, then it doesn't "exist" anymore. Though React can synchronize with new incoming serialized VDOM.

**Reply by surafel-99** [reply]
Thank you! It worked!  I will look into serialized VDOM and the new incoming serialized VDOM. Thank you again!

**Reply by surafel-99** [reply]
Another issue. How to pass the same value to another child server component??

```js
export default function Home() {

  const getToggleValue = async (value) => {
    "use server";
    const childToggleValue = value;       // can I assign the returned value to a variable and pass it to another child component?? 
  }

  return (
    <div className="grid grid-cols-1 md:grid-cols-2  lg:grid-cols-3 justify-center mx-3 my-3 gap-2">
      <Card showDescription = {childToggleValue}/>
    </div>)

}

``` 
### the error message to this one is "Cannot find name 'childToggleValue' " and next.js shows "can't compile message"

**Reply by icyJoseph** [reply]
This is what I meant before. 

The server action is created and the client knows how to call it, but the start happened within the server component and that no longer exists. You are almost trying to replicate a pass callback to set state in the parent pattern, but here you have a component that renders server side, and a client that gets the ability to invoke some server work later on, but the rendering is long gone. 

I think you probably need a context provider, use client, to pass data between these children.

By using composition, client Components can slot in, server components.
